### PR TITLE
Fix logger pickling error

### PR DESCRIPTION
**Pull request by Derek-Wds**
## Description

The previous PR [#393](https://github.com/microsoft/qlib/pull/393) did not solve the logger pickling completely. The `QlibLogger` object cannot be saved either. This PR is trying to fix the problem and make everything work.

## How Has This Been Tested?
- [ ] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [ ] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [x] Fix bugs
- [ ] Add new feature
- [ ] Update documentation

### CODE REVIEW

File: qlib/log.py [hunk 1]
```diff
@@ -17,6 +17,7 @@ def __new__(cls, name, bases, dict):
         wrapper_dict = logging.Logger.__dict__.copy()
         wrapper_dict.update(dict)
         wrapper_dict["__doc__"] = logging.Logger.__doc__
+        del wrapper_dict["__reduce__"]  # make Logger object can be pickled
```
Comment (you-n-g) [inline qlib/log.py:20]: I'm little confused about the current implementation.

If the problem is that copying the pickle related magic methods will result in error when pickling, why don't we just exclude them when copying methods dict.

image

File: qlib/log.py [hunk 2]
```diff
@@ -29,6 +31,15 @@ def __init__(self, module_name):
         self.module_name = module_name
         self.level = 0
 
+    def __getstate__(self):
```
Comment (you-n-g) [inline qlib/log.py:34]: Why  do you have to implement these three magic functions? 
Aren't they the default behaviors of an Python Object?

Comment (you-n-g) [reply qlib/log.py:34]: Maybe these functions should also be skipped in the  `MetaLogger.__new__` function.

Comment (Derek-Wds) [reply qlib/log.py:34]: It is necessary to add these two methods in the class. These functions are used by pickle to customized how objects are pickled. If we do not include these in the class, the `__getattr__` method will be called while pickling and results in recursion error.
Screenshot from 2021-04-30 09-31-04

Besides the original dict did not have these two methods either.
Screenshot from 2021-04-30 09-30-44

It is necessary to specify these two methods if we want to pickle the class.

File: qlib/log.py [hunk 3]
```diff
@@ -29,6 +31,9 @@ def __init__(self, module_name):
         self.module_name = module_name
         self.level = 0
 
+    def __reduce__(self):
```
Comment (you-n-g) [inline qlib/log.py:34]: Why is reduce necessary here?

File: qlib/log.py [hunk 4]
```diff
@@ -15,9 +15,11 @@
 class MetaLogger(type):
     def __new__(cls, name, bases, dict):
         wrapper_dict = logging.Logger.__dict__.copy()
-        wrapper_dict.update(dict)
-        wrapper_dict["__doc__"] = logging.Logger.__doc__
-        return type.__new__(cls, name, bases, wrapper_dict)
+        for key in wrapper_dict:
+            if key not in dict and key != "__reduce__":
+                dict[key] = wrapper_dict[key]
+        dict["__doc__"] = logging.Logger.__doc__
```
Comment (you-n-g) [inline qlib/log.py:21]: image

**Pull request merged**